### PR TITLE
Add bunyan as a logging framework and use it in all non-content processes.

### PR DESCRIPTION
**Pull request by Mossop**
Bunyan was mainly chosen in preference to winston because winston doesn't work in a webpacked environment.

The idea is that the first thing a process should do (before it has chance to log anything) is to configure the logging for the process and after that most code can jsut use the default logger. This gives a slightly weird indirection between the default logger and the actual real logger than isn't created until after startup.

The logger is designed to be mostly swappable for console.

UA and content services get logging to file which helps debugging issues on windows.

### CODE REVIEW

**Comment by Mossop** [comment]
This fixes #319.

File: .eslintrc [hunk 1]
```diff
@@ -48,6 +48,7 @@
     "arrow-body-style": ["off"],
     "func-names": ["off"],
     "space-before-function-paren": ["error", "never"],
+    "no-console": ["error"],
```
Comment (victorporof) [inline .eslintrc]: happy happy joy joy

File: app/ui/browser/index.jsx [hunk 2]
```diff
@@ -34,6 +34,9 @@ import * as actions from './actions/main-actions';
 import * as profileDiffs from '../../shared/profile-diffs';
 import BUILD_CONFIG from '../../../build-config';
 import UserAgentClient from '../../shared/user-agent-client';
+import { configureLogging } from '../../shared/logging';
+
+configureLogging('renderer');
```
Comment (victorporof) [inline app/ui/browser/index.jsx]: I think this should go before every other import, yes?

File: build/main.js [hunk 3]
```diff
@@ -20,16 +20,19 @@ const thenify = require('thenify-all').thenify;
 const semver = require('semver');
 const checkDependencies = require('check-dependencies');
 const VALID_NODE_VERSION_RANGE = require('../package.json').engines.node;
+const { configureLogging, logger } = require('../app/shared/logging');
+
+configureLogging('build');
```
Comment (victorporof) [inline build/main.js]: Ditto here?

File: app/main/browser.js [hunk 4]
```diff
@@ -36,6 +36,9 @@ import * as instrument from '../services/instrument';
 import * as spawn from './spawn';
 import * as BW from './browser-window';
 import UserAgentClient from '../shared/user-agent-client';
+import { configureLogging, logger } from '../shared/logging';
+
+configureLogging('main');
```
Comment (victorporof) [inline app/main/browser.js]: Init this before everything else? Not sure, just asking.

File: app/shared/logging.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+// The entry point to the process should configure this logging service
+let rootLogger = null;
+export function configureLogging(processType) {
+  const streams = [{
+    stream: consolePipe,
+    level: 'debug',
+  }];
+
+  if (processType.endsWith('-service')) {
+    const filestream = fs.createWriteStream(`${processType}.log`);
```
Comment (jsantell) [inline app/shared/logging.js]: Where does this get dumped? Directory root?

**Comment by jsantell** [comment]
Maybe we should turn this off (or push to browser console or something) when in production/packaged build?

File: app/shared/logging.js (hunk 5 above)
Comment (Mossop) [inline app/shared/logging.js]: Right now CWD. Suggestions for alternatives welcome

**Comment by Mossop** [comment]
I'm reworking this a little.

**Comment by Mossop** [comment]
Ok, @victorporof or @jsantell whoever wants to look. This adds a logging file per process to set up the default logger. It's a tiny bit of boilerplate but it means you don't have to be careful to init before anything else could potentially do logging. Unfortunately that makes logging in shared code harder, I'm passing the logger through here but I'd love to hear an alternative.

The logging to files only happens in development builds. Still to current working directory, do we have a profile directory yet?

File: app/main/spawn.js [hunk 6]
```diff
@@ -51,7 +52,7 @@ export function startUserAgentService(userAgentClient, options = {}) {
   // Connecting to a client immediately is sometimes not necessary,
   // e.g. when starting this service standalone.
   if (userAgentClient) {
-    userAgentClient.connect({ port, host, version });
+    userAgentClient.connect({ port, host, version, logger });
```
Comment (victorporof) [inline app/main/spawn.js]: Why is the `logger` here passed as a param, instead of importing it where required?

File: app/services/content-service/index.js [hunk 7]
```diff
@@ -12,14 +12,15 @@
 
 import * as contentService from './server';
 import * as endpoints from '../../shared/constants/endpoints';
+import { logger } from './logging';
 
-process.on('uncaughtException', console.error);
-process.on('unhandledRejection', console.error);
+process.on('uncaughtException', (...args) => logger.error(...args));
+process.on('unhandledRejection', (...args) => logger.error(...args));
```
Comment (victorporof) [inline app/services/content-service/index.js]: Why these changes? Do we have to bind `logger` to `error` and other such functions?

File: app/services/content-service/logging.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+/*
```
Comment (victorporof) [inline app/services/content-service/logging.js]: It's a bit strange having a different "logging" file in each part of the app. Looks like most of this code can be reused. Why not do that?

File: app/services/content-service/server.js [hunk 9]
```diff
@@ -34,7 +35,8 @@ function configure(app, router) {
 export async function start() {
   const { setup } = await makeServer(
     endpoints.CONTENT_SERVER_VERSION,
-    endpoints.CONTENT_SERVER_PORT);
+    endpoints.CONTENT_SERVER_PORT,
+    logger);
```
Comment (victorporof) [inline app/services/content-service/server.js]: Ditto here: why pass `logger` in when it can be imported?

File: app/services/user-agent-service/index.js [hunk 10]
```diff
@@ -13,9 +13,10 @@
 import * as userAgentService from './server';
 import { ProfileStorage } from './storage';
 import meta from './meta.json';
+import { logger } from './logging';
 
-process.on('uncaughtException', console.error);
-process.on('unhandledRejection', console.error);
+process.on('uncaughtException', (...args) => logger.error(...args));
+process.on('unhandledRejection', (...args) => logger.error(...args));
```
Comment (victorporof) [inline app/services/user-agent-service/index.js]: Ditto here for the change. Why not bind in the logging module itself, if necessary?

File: app/services/user-agent-service/logging.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
+/*
```
Comment (victorporof) [inline app/services/user-agent-service/logging.js]: Code reuse maybe? :)

File: app/shared/logging.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    stream: serializer,
+    level,
+  };
+}
+
+/**
+ * Sends log messages to the `console` object
+ */
+export function pipeToConsole() {
+  /* eslint no-console: off */
```
Comment (victorporof) [inline app/shared/logging.js]: Either make this file-level explicitly, or turn on "eslint no-console" at the end of this function.

File: app/ui/content/logging.js [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+ specific language governing permissions and limitations under the License.
+ */
+
+import bunyan from 'bunyan';
+import { pipeToConsole } from '../../shared/logging';
+
+export const logger = bunyan.createLogger({
+  name: 'content',
+  streams: [
+    // TODO pipe this to the main process stdout
```
Comment (victorporof) [inline app/ui/content/logging.js]: Does this happen without the PR? If so, it might be a bit sad to have it gone, but not that big of a deal i think, since we'll see log calls in devtools.

**Comment by victorporof** [comment]
@Mossop how about relying on webpack to include the boilerplate, or at least import the right thing for us in every file, instead of having to import it ourselves?

File: app/main/spawn.js (hunk 6 above)
Comment (Mossop) [inline app/main/spawn.js]: user-agent-client is in shared code, it doesn't know which logger to use.

File: app/services/content-service/index.js (hunk 7 above)
Comment (Mossop) [inline app/services/content-service/index.js]: I should double check this, it might have leaked through from another change. If so I guess I can manually bind those.

File: app/services/content-service/logging.js (hunk 8 above)
Comment (Mossop) [inline app/services/content-service/logging.js]: Of the six different process types we have roughly 3 or 4 different logging configurations. Sharing some of the code is possible but most of it already is. I don't see this changing much except in ways that makes the processes differ more.

bunyan also doesn't seem to allow configuring a logger after creation.

File: app/services/content-service/server.js (hunk 9 above)
Comment (Mossop) [inline app/services/content-service/server.js]: Again, makeServer is in shared code.

File: app/ui/content/logging.js (hunk 13 above)
Comment (Mossop) [inline app/ui/content/logging.js]: I have no idea, it doesn't on Windows I think but stdout is almost completely busted on windows, partly why I'm doing this to make it actually possible to fix.

**Comment by Mossop** [comment]
@victorporof I'd love to use webpack if I could figure out how. Unfortunately some of the services use babel and not webpack right now and even then I couldn't figure out how to make webpack do what I wanted. Let's talk in London!

**Comment by Mossop** [comment]
@victorporof I failed to notice the r+ and ended up reworking this so you always import the same module regardless of process (build scripts aside) and a webpack define chooses what streams to use. This means we're now webpacking everything that ships in the app!

https://cdn.meme.am/instances/500x/68853970.jpg

File: app/main/spawn.js [hunk 14]
```diff
@@ -26,7 +26,29 @@ const CONTENT_SERVICE_PATH = path.join(SERVICES_PATH, 'content-service', 'index.
 // @TODO Will `node` be accessible via path on users machines that don't
 // have node installed as an engineer?
 
-const children = [];
+const children = new Set();
+function spawnProcess(name, command, args, options) {
```
Comment (victorporof) [inline app/main/spawn.js]: Should maybe make this shared, possibly augmenting the existing spawn function we have in build/utils?

File: app/main/spawn.js [hunk 15]
```diff
@@ -26,7 +26,29 @@ const CONTENT_SERVICE_PATH = path.join(SERVICES_PATH, 'content-service', 'index.
 // @TODO Will `node` be accessible via path on users machines that don't
 // have node installed as an engineer?
 
-const children = [];
+const children = new Set();
```
Comment (victorporof) [inline app/main/spawn.js]: Uber nit: newline.

File: app/main/spawn.js [hunk 16]
```diff
@@ -18,7 +18,7 @@ import BUILD_CONFIG from '../../build-config';
 const ROOT = path.join(__dirname, '..', '..');
 const DB_PATH = ROOT;
 const SERVICES_PATH = path.join(ROOT, 'lib', 'services');
-const UA_SERVICE_BIN = path.join(SERVICES_PATH, 'user-agent-service', 'bin', 'user-agent-service');
+const UA_SERVICE_BIN = path.join(SERVICES_PATH, 'user-agent-service', 'user-agent-service');
```
Comment (victorporof) [inline app/main/spawn.js]: Why this change?

File: build/task-build-shared.js [hunk 17]
```diff
@@ -1,14 +0,0 @@
...
-import { SRC_DIR } from './const';
-import { babelBuild, babelFile, watchSource } from './utils';
-
-const source = path.join(SRC_DIR, 'shared');
-
-export default async function() {
-  console.log('Building shared...');
-  await babelBuild(source);
-  return watchSource(source, babelFile);
-}
```
Comment (victorporof) [inline build/task-build-shared.js]: We'll see who lands our PR first. Losers rebase.

File: build/webpack.config.content-service.default.js [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+  output: {
+    path: dest,
+    filename: 'index.js',
+    sourceMapFilename: 'index.map',
+  },
+  target: 'node',
+  plugins: [
+    ...defaultConfig.plugins,
+    new webpack.DefinePlugin({
+      LIBDIR: 'require("path").join(__dirname, "..", "..")',
```
Comment (victorporof) [inline build/webpack.config.content-service.default.js]: Instead of __dirname, can we use a predefined const from const.js?

File: app/main/spawn.js (hunk 16 above)
Comment (Mossop) [inline app/main/spawn.js]: That's the target file for webpack, I could still keep the bin directory around but it seems unnecessary once everything else is gone.
